Src/ILGPU/IR/Transformations/AcceleratorSpecializer.cs

```diff
@@ -382,7 +437,47 @@ static AcceleratorSpecializer()
             WriteToOutput writeToOutput) =>
             context.Remove(writeToOutput);
 
+
+        /// <summary>
+        /// Rebuilds the list of affected blocks.
+        /// </summary>
+        /// <param name="context">The current rewriter context.</param>
+        /// <param name="builder">The current method builder.</param>
+        /// <param name="affectedBlocks">The blocks to rebuild.</param>
+        private static void RebuildAffectedBlocks(
+            IRContext context,
+            Method.Builder builder,
+            HashSet<BasicBlock> affectedBlocks)
+        {
+            // We change the control-flow structure during the transformation but
+            // need to get information about previous predecessors and successors
+            builder.AcceptControlFlowUpdates(accept: true);
+
+            // Build an identity mapping
+            var parameterValues = InlineList<ValueReference>.Create(builder.NumParams);
+            for (int i = 0, e = builder.NumParams; i < e; ++i)
+                parameterValues.Add(builder[i]);
+
+            // Rebuild the affected blocks, forcing constants to be folded
+            var predicate = IsInCollectionPredicate.ToPredicate(affectedBlocks);
+            var partialMode = new IRRebuilder.PartialInlineMode<
+                IsInCollectionPredicate<HashSet<BasicBlock>>>(predicate);
+            var rebuilder = builder.CreateRebuilder(
+                builder.Method.CreateParameterMapping(parameterValues),
+                default,
+                partialMode,
+                builder.SourceBlocks);
+
+            // Create an exit return
+            var (exitBlock, exitValue) = rebuilder.Rebuild();
+            exitBlock.CreateReturn(exitValue.Location, exitValue);
```

Comment:
I think this is correct since we are rebuilding all `affected` blocks. In other words, if we encounter an exit block, this block can perform a `return` operation only. All other blocks should have already been rewired to their successor blocks using  `Branch` terminator. Am I right?

Reply:
Actually, I was not sure if this was always required. The original method should already have a an exit block and return statement, so is it really required? I copied this code from the constant propagation branch.